In [1]:
import pandas as pd
df=pd.read_excel("/content/drive/MyDrive/DS assigment/Association Rules/Online retail.xlsx", header=None)
df

0
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

In [2]:
df.head()

0
0  shrimp,almonds,avocado,vegetables mix,green gr...
1                             burgers,meatballs,eggs
2                                            chutney
3                                     turkey,avocado
4  mineral water,milk,energy bar,whole wheat rice...

In [3]:
df.isnull().sum()

0    0
dtype: int64

In [4]:
df.duplicated().sum()

np.int64(2325)

In [5]:
df[df.duplicated()]

0
34                cookies
42              spaghetti
60              spaghetti
64            turkey,eggs
65           french fries
...                   ...
7491        herb & pepper
7492   chocolate,escalope
7495  pancakes,light mayo
7498              chicken
7499   escalope,green tea

[2325 rows x 1 columns]

In [6]:
df=df.drop_duplicates()

In [7]:
df.duplicated().sum()

np.int64(0)

In [8]:
df.reset_index(drop=True, inplace=True)

In [9]:
df

0
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
5171  burgers,salmon,pancakes,french fries,frozen sm...
5172  turkey,burgers,dessert wine,shrimp,pasta,tomat...
5173                      butter,light mayo,fresh bread
5174  burgers,frozen vegetables,eggs,french fries,ma...
5175    eggs,frozen smoothie,yogurt cake,low fat yogurt

[5176 rows x 1 columns]

In [10]:
transactions=df[0].str.split(',')
transactions

0       [shrimp, almonds, avocado, vegetables mix, gre...
1                              [burgers, meatballs, eggs]
2                                               [chutney]
3                                       [turkey, avocado]
4       [mineral water, milk, energy bar, whole wheat ...
                              ...                        
5171    [burgers, salmon, pancakes, french fries, froz...
5172    [turkey, burgers, dessert wine, shrimp, pasta,...
5173                    [butter, light mayo, fresh bread]
5174    [burgers, frozen vegetables, eggs, french frie...
5175    [eggs, frozen smoothie, yogurt cake, low fat y...
Name: 0, Length: 5176, dtype: object

In [11]:
transactions=transactions.tolist()
transactions[:5]

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea']]

In [12]:
!pip install mlxtend

In [13]:
from mlxtend.preprocessing import TransactionEncoder
encoder=TransactionEncoder()
encoder_a=encoder.fit(transactions).transform(transactions)

In [14]:
df=pd.DataFrame(encoder_a, columns=encoder.columns_)
df.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False     True               True      False     True        False   
1       False    False              False      False    False        False   
2       False    False              False      False    False        False   
3       False    False              False      False     True        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False            True   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...   False           False   
3  False           False      False        False  ...    True           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False              True              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False   True        False     False  
1             False  False        False     False  
2             False  False        False     False  
3             False  False        False     False  
4              True  False        False     False  

[5 rows x 120 columns]

In [15]:
for i in df.columns:
  print(df[i].value_counts(), "\n")

 asparagus
False    5175
True        1
Name: count, dtype: int64 

almonds
False    5024
True      152
Name: count, dtype: int64 

antioxydant juice
False    5118
True       58
Name: count, dtype: int64 

asparagus
False    5142
True       34
Name: count, dtype: int64 

avocado
False    4938
True      238
Name: count, dtype: int64 

babies food
False    5145
True       31
Name: count, dtype: int64 

bacon
False    5111
True       65
Name: count, dtype: int64 

barbecue sauce
False    5096
True       80
Name: count, dtype: int64 

black tea
False    5070
True      106
Name: count, dtype: int64 

blueberries
False    5108
True       68
Name: count, dtype: int64 

body spray
False    5092
True       84
Name: count, dtype: int64 

bramble
False    5163
True       13
Name: count, dtype: int64 

brownies
False    4943
True      233
Name: count, dtype: int64 

bug spray
False    5112
True       64
Name: count, dtype: int64 

burger sauce
False    5132
True       44
Name: count, dtype: int64 


In [16]:
from mlxtend.frequent_patterns import apriori
frequent_items=apriori(df,min_support=0.01, use_colnames=True)

In [17]:
frequent_items

support                               itemsets
0    0.029366                              (almonds)
1    0.011206                    (antioxydant juice)
2    0.045981                              (avocado)
3    0.012558                                (bacon)
4    0.015456                       (barbecue sauce)
..        ...                                    ...
433  0.014683  (mineral water, spaghetti, olive oil)
434  0.016615   (pancakes, mineral water, spaghetti)
435  0.012365     (mineral water, shrimp, spaghetti)
436  0.010819       (mineral water, spaghetti, soup)
437  0.013524   (mineral water, spaghetti, tomatoes)

[438 rows x 2 columns]

In [18]:
from mlxtend.frequent_patterns import association_rules
rules=association_rules(frequent_items, metric='confidence', min_threshold=0.5)

In [19]:
rules

antecedents      consequents  antecedent support  \
0              (chocolate, chicken)  (mineral water)            0.021252   
1            (chocolate, olive oil)  (mineral water)            0.023570   
2               (ground beef, eggs)  (mineral water)            0.028787   
3  (ground beef, frozen vegetables)  (mineral water)            0.024536   
4  (ground beef, frozen vegetables)      (spaghetti)            0.024536   
5               (ground beef, milk)  (mineral water)            0.031685   
6           (pancakes, ground beef)  (mineral water)            0.020866   
7                 (olive oil, milk)  (mineral water)            0.024150   
8                      (soup, milk)  (mineral water)            0.021445   
9                 (soup, spaghetti)  (mineral water)            0.020672   

   consequent support   support  confidence      lift  representativity  \
0            0.299845  0.011012    0.518182  1.728163               1.0   
1            0.299845  0.011978    0.508197  1.694862               1.0   
2            0.299845  0.014490    0.503356  1.678717               1.0   
3            0.299845  0.013331    0.543307  1.811957               1.0   
4            0.229521  0.012558    0.511811  2.229911               1.0   
5            0.299845  0.016036    0.506098  1.687861               1.0   
6            0.299845  0.010819    0.518519  1.729286               1.0   
7            0.299845  0.012365    0.512000  1.707546               1.0   
8            0.299845  0.012365    0.576577  1.922913               1.0   
9            0.299845  0.010819    0.523364  1.745448               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0  0.004640    1.453151       0.430500  0.035514   0.311840    0.277454  
1  0.004911    1.423648       0.419878  0.038462   0.297579    0.274073  
2  0.005858    1.409771       0.416291  0.046125   0.290665    0.275840  
3  0.005974    1.533097       0.459382  0.042857   0.347726    0.293883  
4  0.006926    1.578240       0.565425  0.052000   0.366383    0.283262  
5  0.006535    1.417597       0.420869  0.050827   0.294581    0.279788  
6  0.004563    1.454167       0.430714  0.034913   0.312321    0.277300  
7  0.005124    1.434743       0.424619  0.039678   0.303011    0.276619  
8  0.005935    1.653557       0.490474  0.040025   0.395243    0.308907  
9  0.004621    1.468952       0.436096  0.034934   0.319242    0.279723

In [20]:
rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']].head()

antecedents      consequents   support  confidence  \
0              (chocolate, chicken)  (mineral water)  0.011012    0.518182   
1            (chocolate, olive oil)  (mineral water)  0.011978    0.508197   
2               (ground beef, eggs)  (mineral water)  0.014490    0.503356   
3  (ground beef, frozen vegetables)  (mineral water)  0.013331    0.543307   
4  (ground beef, frozen vegetables)      (spaghetti)  0.012558    0.511811   

       lift  
0  1.728163  
1  1.694862  
2  1.678717  
3  1.811957  
4  2.229911

In [21]:
rules=rules[rules['lift']>1].sort_values(by='lift', ascending=False)

In [22]:
rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']].head()

antecedents      consequents   support  confidence  \
4  (ground beef, frozen vegetables)      (spaghetti)  0.012558    0.511811   
8                      (soup, milk)  (mineral water)  0.012365    0.576577   
3  (ground beef, frozen vegetables)  (mineral water)  0.013331    0.543307   
9                 (soup, spaghetti)  (mineral water)  0.010819    0.523364   
6           (pancakes, ground beef)  (mineral water)  0.010819    0.518519   

       lift  
4  2.229911  
8  1.922913  
3  1.811957  
9  1.745448  
6  1.729286

In [23]:
for i, r in rules.head(10).iterrows():
  antecedents=list(r['antecedents'])
  consequents=list(r['consequents'])
  print(f"Rule: {antecedents} --> {consequents}")
  print(f"Support: {r['support']:.4f}")
  print(f"Confidence: {r['confidence']:.4f}")
  print(f"Lift: {r['lift']:.4f}\n")

Rule: ['ground beef', 'frozen vegetables'] --> ['spaghetti']
Support: 0.0126
Confidence: 0.5118
Lift: 2.2299

Rule: ['soup', 'milk'] --> ['mineral water']
Support: 0.0124
Confidence: 0.5766
Lift: 1.9229

Rule: ['ground beef', 'frozen vegetables'] --> ['mineral water']
Support: 0.0133
Confidence: 0.5433
Lift: 1.8120

Rule: ['soup', 'spaghetti'] --> ['mineral water']
Support: 0.0108
Confidence: 0.5234
Lift: 1.7454

Rule: ['pancakes', 'ground beef'] --> ['mineral water']
Support: 0.0108
Confidence: 0.5185
Lift: 1.7293

Rule: ['chocolate', 'chicken'] --> ['mineral water']
Support: 0.0110
Confidence: 0.5182
Lift: 1.7282

Rule: ['olive oil', 'milk'] --> ['mineral water']
Support: 0.0124
Confidence: 0.5120
Lift: 1.7075

Rule: ['chocolate', 'olive oil'] --> ['mineral water']
Support: 0.0120
Confidence: 0.5082
Lift: 1.6949

Rule: ['ground beef', 'milk'] --> ['mineral water']
Support: 0.0160
Confidence: 0.5061
Lift: 1.6879

Rule: ['ground beef', 'eggs'] --> ['mineral water']
Support: 0.0145
Conf

_____
## Analysis and Interpretation

#### Analyzing the Generated Rules
After running the Apriori algorithm on the "Online retail.xlsx" dataset with 7501 transactions, I generated association rules and filtered them to keep only those with a lift greater than 1. This means the items in these rules are bought together more often than if they were chosen randomly. I sorted the rules by lift to focus on the strongest relationships. The key metrics—support, confidence, and lift—helped me understand how often items appear together, how reliable the rules are, and how strong the connections are compared to chance.

The top 10 rules (from the output in cell 23) are shown below with their metrics:

1. **['frozen vegetables', 'ground beef'] → ['spaghetti']**: Support = 0.0126, Confidence = 0.5118, Lift = 2.2299
2. **['milk', 'soup'] → ['mineral water']**: Support = 0.0124, Confidence = 0.5766, Lift = 1.9229
3. **['frozen vegetables', 'ground beef'] → ['mineral water']**: Support = 0.0133, Confidence = 0.5433, Lift = 1.8120
4. **['spaghetti', 'soup'] → ['mineral water']**: Support = 0.0108, Confidence = 0.5234, Lift = 1.7454
5. **['ground beef', 'pancakes'] → ['mineral water']**: Support = 0.0108, Confidence = 0.5185, Lift = 1.7293
6. **['chicken', 'chocolate'] → ['mineral water']**: Support = 0.0110, Confidence = 0.5182, Lift = 1.7282
7. **['olive oil', 'milk'] → ['mineral water']**: Support = 0.0124, Confidence = 0.5120, Lift = 1.7075
8. **['olive oil', 'chocolate'] → ['mineral water']**: Support = 0.0120, Confidence = 0.5082, Lift = 1.6949
9. **['milk', 'ground beef'] → ['mineral water']**: Support = 0.0160, Confidence = 0.5061, Lift = 1.6879
10. **['eggs', 'ground beef'] → ['mineral water']**: Support = 0.0145, Confidence = 0.5034, Lift = 1.6787

- **Support**: Shows how common the combination is (e.g., 0.0160 means 1.6% of all transactions have both milk and ground beef with mineral water).
- **Confidence**: Shows how often the "then" item is bought when the "if" items are (e.g., 57.66% of transactions with milk and soup also have mineral water).
- **Lift**: Shows how much more likely the items are bought together compared to random (e.g., lift of 2.2299 means frozen vegetables and ground beef with spaghetti is 2.23 times more likely than if they were unrelated).

#### Interesting Patterns and Relationships
From the top rules, I noticed a few clear patterns:

1. **Ground Beef as a Key Ingredient**:
   - Ground beef appears in four of the top 10 rules (with frozen vegetables, pancakes, milk, and eggs), often leading to mineral water or spaghetti. For example, the rule ['frozen vegetables', 'ground beef'] → ['spaghetti'] has the highest lift (2.2299), meaning these items are strongly linked. Another rule, ['eggs', 'ground beef'] → ['mineral water'], has a decent lift (1.6787). This suggests ground beef is a popular base ingredient that pairs with other foods and drinks.

2. **Mineral Water as a Common Add-On**:
   - Mineral water is the consequent (the "then" part) in nine of the top 10 rules. It pairs with combinations like milk and soup (lift = 1.9229), chicken and chocolate (lift = 1.7282), and olive oil and milk (lift = 1.7075). Its frequent appearance with high confidence (around 50% or more) and lift above 1 shows it’s a staple item customers add to many purchases.

3. **Meal Combinations**:
   - The rule ['frozen vegetables', 'ground beef'] → ['spaghetti'] (lift = 2.2299) points to a classic meal pattern: ingredients for a dish like spaghetti with meat sauce. Similarly, ['milk', 'soup'] → ['mineral water'] (lift = 1.9229) could indicate soup preparation (milk as an ingredient) paired with a drink. These rules suggest customers are buying items for complete meals.

4. **Unexpected Pairings**:
   - Some rules, like ['chicken', 'chocolate'] → ['mineral water'] (lift = 1.7282) or ['olive oil', 'chocolate'] → ['mineral water'] (lift = 1.6949), are less obvious. Chicken and chocolate or olive oil and chocolate don’t seem like natural meal pairs, but their link to mineral water suggests customers might be grabbing a drink with diverse shopping baskets.

#### Insights into Customer Purchasing Behavior
Based on these patterns, here are some insights into how customers shop:

1. **Meal Planning**:
   - Customers often buy ingredients for specific meals. The strong link between frozen vegetables, ground beef, and spaghetti (lift = 2.2299) shows they’re likely making dishes like spaghetti bolognese. Stores could bundle these items or place them closer together to encourage more sales.

2. **Mineral Water Popularity**:
   - Mineral water’s dominance in the rules (e.g., confidence of 57.66% with milk and soup) suggests it’s a go-to beverage for many customers, no matter what else they’re buying. This could mean it’s a habit or a healthy choice they add to every trip. Retailers might promote it as a complementary item with meal ingredients.

3. **Versatile Ingredients**:
   - Ground beef’s frequent appearance (e.g., with eggs, pancakes, or milk, all leading to mineral water) shows it’s a versatile ingredient in many customers’ diets. They might use it for burgers, pasta, or breakfast dishes. Offering ground beef promotions with related items (like eggs or spaghetti) could boost sales.

4. **Mix of Healthy and Indulgent Choices**:
   - Rules like ['olive oil', 'milk'] → ['mineral water'] (lift = 1.7075) suggest healthier shopping (olive oil and milk are cooking staples), while ['chicken', 'chocolate'] → ['mineral water'] mixes a protein with a treat. This shows customers balance nutrition and indulgence, often rounding out their carts with mineral water.

#### Conclusion
The association rules reveal that customers tend to shop for meal components (like ground beef and spaghetti) and frequently add mineral water as a staple. The high lift values, especially for rules involving ground beef and spaghetti or milk and soup, highlight strong product relationships that aren’t random. These insights could help the store improve product placement (e.g., putting mineral water near ground beef or soup), create bundle deals, or target promotions to match these buying habits. Overall, the rules show a mix of practical meal planning and consistent beverage preferences in customer behavior.

---


---

## Interpret the Results and Provide Insights into Customer Purchasing Behaviour

The association rules from the "Online retail.xlsx" dataset reveal clear patterns in how customers shop. Based on the discovered rules, here’s what they tell us about purchasing behavior:

1. **Meal Preparation Drives Purchases**:
   - The rule `{frozen vegetables, ground beef} → {spaghetti}` shows customers often buy these items together. This suggests they’re planning meals like spaghetti with meat sauce, combining a protein (ground beef), a vegetable, and a staple (spaghetti). It’s a sign of intentional cooking at home, likely for family or quick dinners.

2. **Mineral Water as a Go-To Add-On**:
   - Rules like `{milk, soup} → {mineral water}`, `{chicken, chocolate} → {mineral water}`, and `{olive oil, milk} → {mineral water}` highlight mineral water’s frequent appearance. Customers add it to their carts no matter what else they’re buying—whether it’s soup ingredients or a mix of chicken and chocolate. This points to a habit of grabbing a drink, possibly for hydration or health reasons, alongside their main items.

3. **Ground Beef’s Versatility**:
   - Ground beef shows up in multiple rules, such as `{eggs, ground beef} → {mineral water}` and `{milk, ground beef} → {mineral water}`. This suggests customers use it in various ways—maybe for breakfast (with eggs), pasta (with spaghetti), or other dishes (with milk). It’s a flexible ingredient in their diets, often paired with a drink like mineral water.

4. **Mix of Practical and Surprising Choices**:
   - Rules like `{chicken, chocolate} → {mineral water}` and `{olive oil, chocolate} → {mineral water}` are unexpected. Chicken and chocolate or olive oil and chocolate don’t naturally go together, but customers still buy them with mineral water. This could mean they’re shopping for different needs in one trip—protein for meals, chocolate as a treat—while adding a consistent beverage.

5. **Health and Convenience Blend**:
   - Combinations like `{olive oil, milk} → {mineral water}` or `{frozen vegetables, ground beef} → {mineral water}` suggest a balance. Olive oil and frozen vegetables hint at healthier cooking, while ground beef and milk are everyday staples. Customers seem to mix nutritious choices with convenient, familiar items, often rounding out their baskets with mineral water.

#### Insights into Customer Purchasing Behaviour
- **Home Cooking Focus**: Customers are buying ingredients (ground beef, spaghetti, eggs) for complete meals, showing they cook regularly and plan ahead. Stores could offer recipe ideas or meal kits to match this habit.
- **Staple Beverage Preference**: Mineral water’s constant presence suggests it’s a must-have for many, whether they’re cooking or just snacking. Retailers might boost sales by placing it near popular items like milk or ground beef.
- **Flexible Shopping Trips**: The variety—health-focused items (olive oil) with treats (chocolate)—shows customers shop for multiple purposes at once. Promotions could target this by bundling meal essentials with small indulgences.
- **Practical Shoppers**: Frequent pairings like milk, eggs, and ground beef with mineral water indicate practical, routine purchases. Stores can make these items easy to grab together to encourage bigger carts.

In short, customers are practical, meal-focused shoppers who value staples like mineral water and versatile ingredients like ground beef, blending health and convenience in their buying habits.

---


___
# **Interview Questions**

## **1. What is Lift, and why is it important?**
- Lift measures the **strength of association** between items.
- **Formula:**  
  $$
  \text{Lift} = \frac{\text{Confidence}}{\text{Expected Confidence (if independent)}}
  $$
- Interpretation:
  - **Lift > 1** → Items occur together **more than expected** (strong association).
  - **Lift = 1** → Items occur **independently**.
  - **Lift < 1** → Items are **negatively associated**.

## **2. What are Support and Confidence?**
- **Support**: How frequently an itemset appears in transactions.
  $$
  \text{Support} = \frac{\text{Transactions containing item A and B}}{\text{Total Transactions}}
  $$
- **Confidence**: How often **B is bought when A is bought**.
  $$
  \text{Confidence} = \frac{\text{Support(A ∪ B)}}{\text{Support(A)}}
  $$

## **3. What are some limitations or challenges of Association rules mining?**
- Requires **large datasets** for meaningful insights.
- Choosing the **right support and confidence thresholds** is tricky.
- Can generate **too many rules**, making it hard to interpret results.
